### PR TITLE
Add more of the new features added in React 16.6

### PR DESCRIPTION
**Pull request by Jessidhia**
Continues off #29990.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://reactjs.org/blog/2018/10/23/react-v-16-6.html
- [x] Increase the version number in the header if appropriate.

### CODE REVIEW

File: types/react/index.d.ts
Comment (Kovensky) [inline types/react/index.d.ts]: I'm not really happy about this. Having an underscored value here looks like reaching into internals; but it's the only way to have access to `T` in `LibraryManagedProps`.

An alternative is to make up our own fictional prop name.

File: types/react/test/managedAttributes.tsx
Comment (Kovensky) [inline types/react/test/managedAttributes.tsx]: I fixed lints that were hit while I was editing this uncommented.

File: types/react/test/managedAttributes.tsx
Comment (Kovensky) [inline types/react/test/managedAttributes.tsx]: Assuming I'm using `$ExpectType` correctly these all pass. It just doesn't _actually_ work when the functional/exotic components are used in JSX.

File: types/react/test/tsx.tsx
Comment (Kovensky) [inline types/react/test/tsx.tsx]: 😇

image

Comment (Kovensky) [reply types/react/test/tsx.tsx]: I couldn't figure out how `ref`s work by reading `mountLazyComponent` in `ReactFiberBeginWork.js`; I actually thought they didn't work at all at first.

I did verify that they work at runtime, though (refs are set/called when the `Suspense` resolves).

**Comment by Kovensky** [comment]
Issue I found while updating a codebase to `React.lazy`:

image

Apparently `Readonly<any>` is actually identical to `{}` and for some reason TS is expecting it to be covariant(?). Removing the `Readonly` is enough to get things to work.

I don't know whether to fix this here or in another PR.

File: types/react/index.d.ts
Comment (Kovensky) [inline types/react/index.d.ts]: React _will_ throw at runtime if the Suspense suspends and the fallback is undefined.

File: types/react/index.d.ts
Comment (artem-malko) [reply types/react/index.d.ts]: Hm, it is strange, cause here https://reactjs.org/docs/context.html#classcontexttype we can see, React allows us to use just this.context, but it is not possible, if context will be deleted from types of Component.

Comment (artem-malko) [reply types/react/index.d.ts]: By the way, there is no anything like value in Component.contextType.Provider.

Comment (Kovensky) [reply types/react/index.d.ts]: The point is to force the user to declare it themselves because there is nothing better we can do other than `any` here, and having it already declared would allow people to access it even without a `static contextType`.

`value` is the `value` prop of the `Provider` exotic component.

Comment (artem-malko) [reply types/react/index.d.ts]: Hm, could you explain, why this code is not working:
```
class Component<P, S> {
    static contextType: Context<any>;
    ...
    ...
    context: typeof Component.contextType extends Context<infer T> ? T : never
}
```

Type of context is any always( But it works perfectly in my own component.

Comment (Kovensky) [reply types/react/index.d.ts]: That is because, by declaring it in the superclass, you don't have access to its type as assigned in the subclass. The type of the constructor itself is not generic so it just uses the type as declared, assuming `Component` is constructed as is and not through a subclass. (Note: despite how I phrased it, this still wouldn't work if `Component` was `abstract`)

This is the deficiency in TypeScript's class declaration language: it is possible to make an instance type generic, but not the constructor object itself generic.

There is a possible hack with constructible interfaces, though, if they're even accepted in an `extends` clause. It could be possible to just completely replace the `class Component` declaration with a pair of `interfaces` and a `const`. But... at what cost?.

Comment (artem-malko) [reply types/react/index.d.ts]: Thank you for explanation. I agree, what it is ok to get rid of context from typings.

**Review by mikiest — APPROVED** [review]
Nice one!
Just a couple of minor comments about possibly avoidable `any`s.
Also, is it worth cleaning up some comments before merging?

File: types/react/index.d.ts
Comment (Kovensky) [reply types/react/index.d.ts]: I kept it the same type as `catch` has.

```ts
try {
  ...
} catch (e) {
  // e will _always_ be `any`
  // giving it another type is a build error
}
```

You can also throw anything in javascript; even `throw undefined`.

File: types/react/index.d.ts
Comment (Kovensky) [reply types/react/index.d.ts]: Yeah, if there is no argument it'll be inferred as `{}` and this will actually cause the compiler to reject subtypes of `ComponentType<{}>` because `{}` will be a function argument.

It's the problem of `(arg: { foo: string }) => null` not being assignable to `(arg: {}) => null`, but being assignable to `(arg: any) => null`.

**Comment by plievone** [comment]
@Kovensky I have used the following React.lazy definition for a while in my types.d.ts:
```ts
interface ModuleLike<T> {
    default: T;
}
import * as React from 'react';
declare module 'react' {
    /**
     * Otherwise works, but 'show all usages' work only for prop names, not for component names, unfortunately
     */
    function lazy<T>(f: () => T): T extends Promise<infer U> ?
        (U extends ModuleLike<infer V> ? V : never) :
        never;
}
```

**Comment by Kovensky** [comment]
I am using this `react16.6.d.ts` at work to at least augment `react` with the new types. This can't change most of the existing types but it at least adds the types for the new functions.

```ts
export {}

declare module 'react' {
  export interface ExoticComponent<P = {}> {
    /**
     * # WARNING
     *
     * Exotic components are not actually callable
     *
     * @deprecated Exotic components are not callable
     */
    (props: P): (ReactElement<any>|null)
    readonly $$typeof: symbol
  }

  export interface NamedExoticComponent<P = {}> extends ExoticComponent<P> {
    displayName?: string
  }

  interface Context<T> {
    displayName?: string
  }

  export type ComponentProps<T extends ComponentType<any>> =
    T extends ComponentType<infer P> ? P : {}
  export type ComponentPropsWithRef<T extends ComponentType<any>> =
    T extends ComponentClass<infer P>
      ? P & ClassAttributes<InstanceType<T>>
      : T extends SFC<infer P>
        ? P
        : {}

  export function memo<P extends object> (
    Component: SFC<P>,
    propsAreEqual?: (prevProps: Readonly<P & { children?: ReactNode }>, nextProps: Readonly<P & { children?: ReactNode }>) => boolean
  ): NamedExoticComponent<P>
  export function memo<T extends ComponentType<any>> (
    Component: T,
    propsAreEqual?: (prevProps: Readonly<ComponentProps<T>>, nextProps: Readonly<ComponentProps<T>>) => boolean
  ): NamedExoticComponent<ComponentPropsWithRef<T>>

  /**
   * This feature is not yet available for server-side rendering.
   * Suspense support will be added in a later release.
   */
  export const Suspense: ExoticComponent<{
    children?: ReactNode

    /** A fallback react tree to show when a Suspense child (like React.lazy) suspends */
    fallback: ReactNode|null

    // I tried looking at the code but I have no idea what it does.
    // https://github.com/facebook/react/issues/13206#issuecomment-432489986
    /**
     * Not implemented yet, requires unstable_ConcurrentMode
     */
    // maxDuration?: number
  }>

  function lazy<T extends ComponentType<any>> (
    factory: () => Promise<{ default: T }>
  ): ExoticComponent<ComponentPropsWithRef<T>>
}
```

**Comment by Kovensky** [comment]
Looks like this still needs an Owner review to get the `Merge: Express` tag 😇

**Comment by lednhatkhanh** [comment]
So excited waiting for this to be merged!

**Comment by johnnyreilly** [comment]
@Kovensky can you resolve the conflicts and I'll merge this? (Please! 😁)

**Comment by Kovensky** [comment]
Rebased :fire:

**Comment by johnnyreilly** [comment]
Nice! Just wait for ci to finish and we'll get this in too!

**Comment by johnnyreilly** [comment]
Done!

**Comment by maciejw** [comment]
if @types/react has version 16.7.1 in npm then it should contain things from react 16.7.0-alpha.0 like maxDuration in React.Suspense witch is np commented out, what do you think?

File: types/react/index.d.ts
Comment (diegohaz) [inline types/react/index.d.ts]: Is this right? If I have a render prop component (where children might be a function), it's not being inferred as functional component.

Comment (Kovensky) [reply types/react/index.d.ts]: Implicit `children` strike back

Again

🙄

I guess I have a New Years break project for a breaking change. Hopefully one I can get to compile.

Comment (diegohaz) [reply types/react/index.d.ts]: I'm investigating it further. The problem may not be `children`, but something else (my current bet is functional components with custom generic types). I'll see if I can come up with a minimum reproduction.

Comment (Kovensky) [reply types/react/index.d.ts]: @sandersn @johnnyreilly Would it be possible to condition this breaking change on a specific TypeScript version through typesVersion?

I was thinking of being able to make the change incremental by only updating types on demand, when people update to, say, TS3.3. This would require kind of a "max supported TS version" on types that depend on `@types/react`.

I'm a bit afraid of it being a python3 problem, though, but the technical debt of `@types/react` is now nigh-unmanageable.

Yes, I could work around this problem in this specific instance by changing the input to a function instead of FunctionComponent, but it's pervarsive and everywhere. A lot of things that currently work in the ecosystem (like function children) work because of bugs in the types, not because of features.

Comment (Kovensky) [reply types/react/index.d.ts]: @diegohaz ah, yes, that is the one big problem with mapped types (and the lack of kinds?); they erase generics and collapse unions. If the props are either of those the output type may be funky.

Comment (ericanderson) [reply types/react/index.d.ts]: In another thread we decided it was time to update the types to expect TS 3

Comment (johnnyreilly) [reply types/react/index.d.ts]: > Would it be possible to condition this breaking change on a specific TypeScript version through typesVersion?

I'm afraid I haven't played with `typesVersion` yet and so can't advise. That said, as @ericanderson mentioned there's been prior discussion about upgrading to TS 3. We should do that anyway. If anyone is thinking about submitting a PR that does that then they should expect ♥️ and 🤗

**Pull request merged**
